### PR TITLE
[vcpkg_download_distfile] Pad the SHA to re-enable testing the hash

### PR DESCRIPTION
**Pull request by ras0219**

### CODE REVIEW

**Review by autoantwort — APPROVED** [review]

**Review by NancyLi1013 — APPROVED** [review]

File: scripts/cmake/vcpkg_download_distfile.cmake [hunk 1]
```diff
@@ -80,6 +80,11 @@ function(vcpkg_download_distfile VAR)
             message(FATAL_ERROR "vcpkg_download_distfile must not be passed both SHA512 and SKIP_SHA512.")
         endif()
     endif()
+    string(LENGTH "${vcpkg_download_distfile_SHA512}" vcpkg_download_distfile_SHA512_length)
+    if(vcpkg_download_distfile_SHA512_length LESS 128)
+        # Pad with zeroes to make a valid SHA512
+        string(SUBSTRING "${vcpkg_download_distfile_SHA512}00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000" 0 128 vcpkg_download_distfile_SHA512)
+    endif()
```
Comment (strega-nil-ms) [inline scripts/cmake/vcpkg_download_distfile.cmake:87]:
```suggestion
    if(vcpkg_download_distfile_SHA512 STREQUAL "0")
        string(REPEAT "0" 128 vcpkg_download_distfile_SHA512)
    endif()
    string(LENGTH "${vcpkg_download_distfile_SHA512}" vcpkg_download_distfile_SHA512_length)
    if(NOT vcpkg_download_distfile_SHA512_length EQUAL 128)
        message(FATAL_ERROR "Invalid SHA512: ${vcpkg_download_distfile_SHA512}")
    endif()
```

can we do something like this instead?

Comment (strega-nil-ms) [reply scripts/cmake/vcpkg_download_distfile.cmake:87]: or maybe:
```suggestion
    string(LENGTH "${vcpkg_download_distfile_SHA512}" vcpkg_download_distfile_SHA512_length)
    if(NOT vcpkg_download_distfile_SHA512_length EQUAL "128")
            string(REPEAT "0" 128 vcpkg_download_distfile_SHA512)
    endif()
```

**Review by strega-nil-ms — APPROVED** [review]

**Comment by autoantwort** [comment]
Yeah but now it doesn't work anymore if I use `1` as SHA, or something else

**Review by vicroms — APPROVED** [review]

**Comment by ras0219-msft** [comment]
That's true, but I think for the purposes of local development it is not a burden to require use of the literal 0 (compared with the burden of synthesizing _exactly_ 128 `0`'s).

**Comment by autoantwort** [comment]
> That's true, but I think for the purposes of local development it is not a burden to require use of the literal 0 (esp. compared with the burden of synthesizing _exactly_ 128 `0`'s). 

The code to fill up the hash with zeros up to a string length of 128 was not that complicated. 

> Most importantly, the error messages should make the path forward clear.

No, it only shows that I have entered an invalid sha, e.g. for the input `1`. Imho not very user friendly. Imho it should just do it like in your first version, it is the easiest one for the users.

**Comment by strega-nil-ms** [comment]
Thanks again robert!

**Comment by autoantwort** [comment]
When you stick with the only `0` way you should update the documentation [here](https://github.com/microsoft/vcpkg/blob/master/docs/examples/packaging-github-repos.md), because there you say that users should use `1`

**Pull request merged**